chore: Readme for upstream edk2-staging (#9)
timberland-sig/edk2: readme.md

# UEFI NVMe-oF Boot Implementation
## Introduction
This UEFI Nvme-oF Boot Implementation enables a Boot-from-SAN prototype that produces and consumes [ACPI](https://uefi.org/specifications) XSDT [NBFT](https://uefi.org/specs/ACPI/6.5/05_ACPI_Software_Programming_Model.html?highlight=nbft#description-header-signatures-for-tables-reserved-by-acpi) descriptors that are conformant to the [NVM Express Boot Specification 1.0](https://nvmexpress.org/specification/nvme-boot-specification/) and it’s interface with the [NVMe-oF messaging type](https://uefi.org/specs/UEFI/2.10/10_Protocols_Device_Path_Protocol.html#nvme-over-fabric-nvme-of-namespace-device-path) to [UEFI System Specification 2.10](https://uefi.org/specs/UEFI/2.10/).
 
Further implementation details relative to the UEFI NvmeOfDxe driver can be found in the [POC Readme](https://github.com/timberland-sig/edk2/blob/timberland_1.0_final/NetworkPkg/NvmeOfDxe/readme.md).
 
## Known Open Issues and Errata
 

> - [ ] IPv6 still has general config and interop issues; primarily:

Reviewer: why did you add check marks `[ ]` here?

Author: To give us a list of what has or hasn't been fixed beyond the changelog.  I'm fine converting to normal bullets if it's too confusing.

Reviewer: If you think it's helpful, leave it in. My €0.02 is that if we fix those issues, we will just modify this README, too.

Author: covered by collapse now

>     - [ ] IPv6 DHCP6 Type 3/4 missing

Reviewer: What do you mean with "DHCP6 Type 3/4"? I am unaware of this terminology.

TBH, I suggest shortening this paragraph to just "IPv6 still has general config and interop issues".

Author: >What do you mean with "DHCP6 Type 3/4"? I am unaware of this terminology.

Apologies, this was in reference to [DUID type 3 and 4](https://www.rfc-editor.org/rfc/rfc8415.html#section-11.1) support.  It should probably say that. 

> TBH, I suggest shortening this paragraph to just "IPv6 still has general config and interop issues".

I'm fine with that if @mrabeda doesn't think the detail is needed here.  We can/should probably make these GH issues anyways (they are over in edk2-private still and need porting in here).

Reviewer: > this was in reference to DUID type 3 and 4 support

I'm confused. I thought we did support DUID-UUID (and I think I argued elsewhere that we don't need DUID-LL and DUID-V, at least it's low prio). Anyway I think this is the least of our problems with IPv6.

Author: The current code supports DUID-EN in the same way that iSCSI and HttpBoot do.  Support for UUID is something tbd.

Author: covered by collapse now

    - [ ] IPv6 RA may still be ongoing during connection attempts
    - [ ] IPv6 SLAAC addresses may not be populated right
    - [ ] DHCP6_OPT_BOOT_FILE_URL broken/poorly tested, needs better string parsing handling

> - [ ] Hii Currently Missing

Reviewer: I suggest adding a note here that the configuration is done using NvmeOfCli.efi instead.

Author: ```suggestion
- [ ] Hii Currently Missing; currently configuration is accomplished using NetworkPkg/Applications/NvmeOfCli.efi
```
Acceptable?

Reviewer: "... which uses a config file in text format" ?

Author: It doesn't have to, you could manually specify some of that stuff - I'm not sure I recommend that for sanity.

> - [ ] Hii Currently Missing

Author: ```suggestion
- [ ] Hii Currently Missing; currently configuration is accomplished using NetworkPkg/Applications/NvmeOfCli.efi. 
```

> - [ ] OVMF/QEMU: Live Network changes can cause hangs in NVMe-oF driver

Reviewer: Do we have an issue for this?

I'm not sure if I understand correctly, but is the FW expected at all to handle "live network changes"? I had this discussion with Hannes recently — can EFI handle e.g. a successful NVMe connection which then breaks down before the boot files (bootloader, kernel, initrd etc) are fully downloaded? Hannes and I thought that this just can't be expected to work.

Author: > Do we have an issue for this?

yes, https://github.com/timberland-sig/edk2-private/issues/76 in the old repo.  I've still not migrated the open issues from there to here yet. On my list. 

>  but is the FW expected at all to handle "live network changes"

Needs to be clarified in the issue - but the issue was that it causes the driver to hang indefinitely needing a qemu reset.  It should have just errored.

Reviewer: > yes, [timberland-sig/edk2-private#76](https://github.com/timberland-sig/edk2-private/issues/76) in the old repo. 

What I reported back then was not a "live" change. It was rather a change applied between reboots, such that the NV data wouldn't match the actual network any more (for example, added or removed vlan tag).

Author: ok, i'll strike this for now.  We're going to find new issues anyways so we can always note errata as they come up or are resolved. 
```suggestion
```

> - [ ] Sightings where EBS handling from EFI_EVENT_GROUP_EXIT_BOOT_SERVICES is touch and go

Reviewer: Not sure what "touch and go" means here.

Author: Perhaps?
```suggestion
- [ ] Sightings where EBS handling from EFI_EVENT_GROUP_EXIT_BOOT_SERVICES is not caught and the NBFT is not installed when a TSL application such as iPXE is also invoked
```

or simply:

```suggestion
- [ ] Sightings where EBS handling from EFI_EVENT_GROUP_EXIT_BOOT_SERVICES is not caught
```

Reviewer: Hm "EBS" stands for "Exit Boot Services", right? "Where Exit Boot Services handling from EFI_EVENT_GROUP_EXIT_BOOT_SERVICES is not caught" doesn't make sense to me. If I understood correctly what @mrabeda said, some EBS handlers are using resources which they aren't supposed to use any more in an EBS handler, and that can cause crashes or hangs.

Author: There is some concern that either we're not setting up `EFI_EVENT_GROUP_EXIT_BOOT_SERVICES` right or something still has a way to avoid the event group callback; but reproduction of this has been iffy.

Reviewer: Let's not go into too much detail then. Say something like "sporadically, issues in the ExitBootServices code path have been observed".

Author: sure
```suggestion
- Sporadically, issues in the ExitBootServices code path have been observed
```

> - [ ] Sightings where EBS handling from EFI_EVENT_GROUP_EXIT_BOOT_SERVICES is touch and go

Author: Perhaps?
```suggestion
- [ ] Sightings where EBS handling from EFI_EVENT_GROUP_EXIT_BOOT_SERVICES is not caught and the NBFT is not installed when a TSL application such as iPXE is also invoked
```

or simply:

```suggestion
- [ ] Sightings where EBS handling from EFI_EVENT_GROUP_EXIT_BOOT_SERVICES is not caught
```
